### DISCUSSION TITLE
constraints on the derivative of control variable

### DISCUSSION THREAD

**Discussion by zmdntnu** [do-mpc/do-mpc#218] [Q&A]
Hi,

One of my control variables denoted as u stands for the rudder angle of a vessel, which should be regulated by the maximum turning rate u_dot_max. So, how should I formulate this in the code? 

Best regards

**Comment by 4flixt** [comment]
Hi @zmdntnu ,

you could introduce an input which is ``u_rudder_dot``, then you introduce a state which is your "real" input ``u_rudder``. This state should consider ``u_rudder_dot`` for its dynamic - either integrate it in continuous time or with the discrete time formulation: ``u_rudder_next = u_rudder + u_rudder_dot``.

After you have done this, you can consider ``u_rudder_dot`` in the cost function or for constraints or anything else ...

Best
4flixt

**Reply by zmdntnu** [reply]
Thank you for your reply. @4flixt 

Do you mean something like the following?

u_rudder= model.set_variable('_u','u_rudder')
u_rudder_dot = model.set_variable('_z','u_rudder_dot')
model.set_rhs('delta',u_rudder_dot )

mpc.bounds['upper','_z','delta_dot'] = u_dot_max
mpc.bounds['lower','_z','delta_dot'] = -u_dot_max


BR

**Reply by zmdntnu** [reply]
delta --> u_rudder

**Reply by zmdntnu** [reply]
I tried, but It seems that we can't use function model.set_rhs() for control variable

**Reply by 4flixt** [reply]
Dear @zmdntnu,

No this is not what I mean. You don't need to use algebraic variables (``type='_z'``). Also please try to understand what right hand side (rhs) means. Obviously, you cannot set a rhs for an input (``type='_u'``) only states (``type='_x'``) can and MUST have a rhs.
Please try to read more about the basics of dynamic systems, control and MPC. I think this will help you to understand my original answer. 

Best,
4flixt
